packages/next-swc/crates/next-core/src/next_shared/transforms/relay.rs

```diff
@@ -0,0 +1,125 @@
+use swc_core::{
+    common::{util::take::Take, FileName},
+    ecma::{
+        ast::{Module, Program},
+        transforms::base::resolver,
+        visit::{FoldWith, VisitMutWith},
+    },
+};
+use swc_relay::RelayLanguageConfig;
+use turbo_binding::{
+    turbo::tasks_fs::FileSystem,
+    turbopack::ecmascript::{
+        CustomTransformer, OptionTransformPluginVc, TransformContext, TransformPluginVc,
+    },
+};
+
+use crate::next_config::{NextConfigVc, RelayConfig, RelayLanguage};
+
+#[derive(Debug)]
+pub struct RelayTransformer {
+    config: swc_relay::Config,
+}
+
+impl RelayTransformer {
+    pub fn new(config: &RelayConfig) -> Self {
+        // [TODO]: There are config mismatches between RelayConfig to swc_relay::Config
+        let swc_relay_config = swc_relay::Config {
+            artifact_directory: config.artifact_directory.as_ref().map(PathBuf::from),
+            language: config.language.as_ref().map_or(
+                RelayLanguageConfig::TypeScript,
+                |v| match v {
+                    RelayLanguage::JavaScript => RelayLanguageConfig::JavaScript,
+                    RelayLanguage::TypeScript => RelayLanguageConfig::TypeScript,
+                    RelayLanguage::Flow => RelayLanguageConfig::Flow,
+                },
+            ),
+            ..Default::default()
+        };
+
+        Self {
+            config: swc_relay_config,
+        }
+    }
+}
+
+#[async_trait]
+impl CustomTransformer for RelayTransformer {
+    async fn transform(
+        &self,
+        program: &mut Program,
+        ctx: &TransformContext<'_>,
+    ) -> Option<Program> {
+        let root = if let Ok(file_path) = ctx.file_path.await {
+            file_path
+                .fs
+                .root()
+                .await
+                .map_or(PathBuf::new(), |v| PathBuf::from(v.path.to_string()))
+        } else {
+            PathBuf::new()
+        };
+
+        let (root, config) = if self.config.artifact_directory.is_some() {
+            (root, None)
+        } else {
+            let config = swc_relay::Config {
+                artifact_directory: Some(PathBuf::from("__generated__")),
+                ..self.config
+            };
+            (PathBuf::from("."), Some(config))
+        };
+
+        let p = std::mem::replace(program, Program::Module(Module::dummy()));
+        *program = p.fold_with(&mut swc_relay::relay(
+            config.as_ref().unwrap_or_else(|| &self.config),
+            FileName::Real(PathBuf::from(ctx.file_name_str)),
+            root,
+            // [TODO]: pages_dir comes through next-swc-loader
+            // https://github.com/vercel/next.js/blob/ea472e8058faea8ebdab2ef6d3aab257a1f0d11c/packages/next/src/build/webpack-config.ts#L792
+            None,
+        ));
+
+        let is_typescript = match &self.config.language {
+            RelayLanguageConfig::TypeScript => true,
+            _ => false,
+        };
+
+        program.visit_mut_with(&mut resolver(
```

Comment:
@kdy1 It doesn't seem like this is the correct way.

I tried to pass `ctx.unresolved_mark` and let it be applied, like

```
let ctxt = SyntaxContext::empty().apply_mark(unresolved_mark);
let ident = Ident::new("require".into(), Span::new(...).with_ctxt(ctxt.unwrap()));
```

But weirdly it gives error

```
    Panic: PanicInfo { payload: Any { .. }, message: Some(assertion failed: `(left != right)`
      left: `Mark(0)`,
     right: `Mark(0)`), location: Location { file: "/Users/ojkwon/.cargo/registry/src/index.crates.io-6f17d22bba15001f/swc_common-0.31.5/src/syntax_pos/hygiene.rs", line: 375, col: 13 }, can_unwind: true }
```

Probably I'm doing something wrong?